### ISSUE TITLE
Locking in GVFS

### ISSUE THREAD

**Issue by Kleptine** [Microsoft/GVFS#110]
I've been working for a while to figure out the best way to support Git for game development. Right now the vast majority of game development studios use Perforce, which is closed-source, poorly implemented, and extremely expensive (causing many studios to use old builds from 2012 rather than upgrade). 

The requirements for game development essentially boil down to two concerns:
 - **Suitable performance for large binary file revisioning**
Most of the assets that are committed on a daily basis are binary files. Images, Music, Animations, Rigs, Models can all easily take up 20-100GB at the head revision for a single game. There's really no way to get around the size of assets here, especially for formats like audio.
 - **File-Locking**
The above formats are all fundamentally un-mergeable data types. ie. There's fundamentally no way you can project a .wav file into a text file and merge it by hand, and there's no way to generate the audio at build time -- it's a static file type. Working under this assumption, the only suitable option for avoiding conflicts (and potentially redoing days of work) is to prevent concurrent file modifications before they happen: File Locking.

GVFS (and GitLFS for the most part) comfortably cover the performance aspect here, so the main missing piece is the file locking behavior. As I see it, a good file-locking solution has a couple of requirements:
 1. **It must integrate well with existing software that may not expect files to be locked.**
Most programs and editors have no concept of file-locking. One main issue with Perforce is that it requires programs to add custom support for Locking/Unlocking a file on write. This leads to shoddily implemented source control plugins, and strictly limits the set of tools that can be used with file locking. 
 2. **It must be total, or as close to total as possible.**
The system should strive to prevent as many concurrent modification issues as possible. For example, if two users simultaneously add file `repo/fileA.bin`, this should be prevented, or warned. 
 3. **It must support normal git workflow.** 
Locking should be implemented in a git-friendly methodology. Ideally, a code developer that doesn't touch binaries files wouldn't even notice the lock system's presence. Locking should not assume a "master" branch and generally should work at the commit object level. (ie. a file can only be locked if you have checked out a descendant commit of the most recent modification). Locking should play nicely with normal git actions like reset, merge, branching, etc.
 4. **It must be straightforward enough for non-technical workers to manage.** 
It must be simple enough to provide a clean UI to artists, or be invisible enough that it doesn't take much understanding of git. Ideally, it is transparent enough to simply use SourceTree or other git visual clients.

### Locking at the File System Level
I've thought a lot about the level at which a locking system should exist. Right now most locking systems operate at the user-level, marking files read-only, and requiring editors to manually unlock files before writing (see P4VS). They fail points 1 and 2. In order to satisfy points 1 and 2 above, I believe locking needs to act at the File System level.

Fundamentally, locking is a constraint on the file system, and not a separate tooling on top of it.

I think an ideal system for locking is where when opening a file, the file system would automatically lock the path for writes, similar to the lazy loading of files on a network drive. Given no lock contention, a program would operate on files as normal, without needing knowledge of the underlying locking system. Given contention, programs would interact via the normal file system errors that already exist for writing to non-writable files. Further, if we want to support the case in point 2, the file system is really the only place this can happen. It would need to recognize when new files are created and lock on the specific file path.

### GVFS

Presumably, this could be implemented as an additional filter after GVFS (although I'm not particularly familiar with file system drivers). GVFS is already integrating with both the file-system and git, though, which makes it a fairly comfortable spot to place this functionality. Additionally, I believe file locking may require a couple of git modifications, which would need to be merged with the modifications made for GVFS.

I've only skimmed the GVFS source, so my understanding of the changes needed is limited, however I think roughly it needs a couple of things:

- A shared ledger of acquired locks. An append-only list of the path locks taken and released and which commits they are taken at. 
- Hooks at the file system level to:
 - Write to this ledger on file modification. (Lock a file on write, create, or delete)
 - Reject writes, create, delete operations on lockable files, unless the HEAD commit is a descendant of the most recent commit on the ledger for that file.

I'd love to hear more thoughts on locking in general. My goal is to achieve a file locking workflow that plays as nicely with git as possible. Happy to provide more context on the reasoning / workflow at our studio, too. Once I can get the GVFS source building, I'd like to start poking around and see what this might actually entail regarding the design of GVFS.

**Comment by sanoursa** [comment]
This is a really great write up, thank you! I've followed through this same reasoning in the past and land on very similar thoughts, but I've also poked some holes in it that I haven't found great solutions for. What are your thoughts on the following:

* Totality

I completely agree with you that locking is only effective if it's reliable and complete. There's nothing more frustrating than being told you can make your edits, spend hours on them, and then have to throw your work away because someone else beat you to it.

However, GVFS does not control the entry points into your repo. Ways in which you could get around any locking in GVFS:
1. Clone the repo with git and make whatever changes you want. It'll take (much) longer to clone and run git commands, but once you're past that hurdle, you can avoid dealing with the lock.
2. Clone your repo with GVFS, hydrate your files, disable GVFS - and make whatever edits you want. 
3. Edit / upload files via the web. If you're using any git service that allows web edits, that's also a very easy way to get around the locks

For that reason, I keep coming to the conclusion that the only place that you can _enforce_ a lock is at push time on the server. That would cover GVFS clone, git clone, and web edits, but it has a lot of implementation drawbacks as well - e.g. push-time policies are very expensive to implement.

* Best effort notifications

If we accept that locks can only be enforced at push time, that's safe and correct, and very inconvenient for the artists who still end up wasting their time. So in addition to enforcement, we need best-effort notification of whether your edits will be successful or not. Your proposal for using GVFS to block edits if you can't get a lock certainly helps there, but you would also need modifications to the web UI of whatever service you're on... and maybe some way to do it in plain old git? Or just make the simplifying assumption that no one will be willing to wait hours for a clone just to get around a lock that they can't push to the server anyway.

* Rebase

What happens if I'm currently at a commit that is a descendant of the most recent modification, I get the lock and make changes, and then I rebase my changes onto another commit that should not have been able to get the lock?

* Branches

I like your idea that you can only get the lock if you're currently at a commit that is a descendant of the last modification. However, this assumes that all changes eventually flow back into master and that there are no other long lived branches. What if your workflow is that you create a release branch for every release, and only commit hotfixes into that branch moving forward, while master carries on? Now you need a notion of a lock per file per branch

* Good UX

Yes we can block writes from the file system. But how will an artist know that the write is failing because they don't have the lock? And what if someone opens a file and works for hours without hitting Ctrl+S (gasp!) only to find that they can't get the lock?

* Lots of questions about the "lock server"

These are mostly design/implementation details that can be worked out once you get past the issues above.

  * Who holds the master information about the lock state of each file? 
  * What happens if a user obtains the lock and then never releases it? We could time out, but what if the timeout is 2 hours and some artist took 2 hours and 5 minutes to make their final edits? Do we need to capture every write event and let the server know we're still busy? What will that do to the perf of our file system?
  * What happens if someone wants to work offline and they're willing to lose their work but the file system still won't let them make edits because it can't reach the lock server?


Overall, I do think file locking could be made to work, but I think GVFS would play only a small part in a bigger solution.

**Comment by derrickstolee** [comment]
I'm familiar with the way that Git LFS does locking, and I know it to be very inadequate. The essential design is that some central server keeps track of a list of locks, and the repo tracks certain file types that are subject to locks (in the `.gitattributes` file). If everyone has the right version of Git LFS installed, then it marks the files as read-only until the lock is acquired from the central server. This requires each user uses the following pattern:

1. Acquire a lock.
2. `git pull origin master`
3. Make changes.
4. Push and complete PR into master.
5. Unlock.

If any user doesn't follow this _very carefully_ they risk having their work be wasted. Not to mention this doesn't stop users from using File Explorer to give themselves write permissions. (GVFS may be able to prevent this at the driver level, giving a slight bit of extra behavior over Git LFS.)

As soon as you think of using any multiple-branch workflow, this all goes out the window and becomes much more difficult. There is no way to lock a file according to a given target branch. Thus, users working against the trunk block users trying to fix a maintenance branch.

Before considering a locking proposal, we need to learn from this existing approach and be sure we are improving on that behavior.

**Comment by Kleptine** [comment]
Hey all,

I've spent a while developing a reasonable system for locking and I think I've come to something that I like, and is quite total (with a couple of restrictions). I'd love feedback or counter-examples where the system would do something unexpected. 

Currently I've been focusing on building a system that works for a single commit graph. However, if you imagine the entire set of local repo's merged together as a single commit graph (just shared across devices), you can see how this might work, given the proper synchronization and networking.

## Overview

The thing that fixed most of the issues of the previous systems is to realize that extra information is necessary to properly handle branching. As an example, there is no way to differentiate between a topic branch and a maintenance branch in raw git. The only difference is the _conventions_ on how they are used. 

![image](https://user-images.githubusercontent.com/1078792/42396055-1067f7c4-8115-11e8-8e83-022cf0396899.png)

For instance, there is a convention that a later topic branch will never merge into maintenance (if it did, it would merge in changes from master as well, which is not-intended).

The problem with this in a locking system is that we need these two branches to act differently, proactively. **Two topic branches shouldn't be able to modify the same file in parallel, while a topic and a maintenance branch should be able to modify files in parallel.**

So we need to declare these branch relationships ahead of time. The simplest way to represent this is to declare which branches "are intended to merge into another". This maps nicely because conflicts mainly occur during merge commits with other branches. We can represent this as arrows in a graph between branches. See this simple example:

![image](https://user-images.githubusercontent.com/1078792/42396479-b1cb4a02-8116-11e8-8b8e-0d99b591163c.png)

I'll skip over it for simplicity, but we can make reasonable assumptions about default edges to add in the graph in most cases. We wouldn't need to do any extra setup for simple topic branches, only when we wanted to represent a more complicated workflow.

## Rules

When deciding if we can make a change to a file on a branch, we are mainly interested in the set of 'conflicting branches', or the set of other branches which might merge from/into this branch. The above dependency graph can be traversed to generate this list of branches.

The fundamental rule is the following, and is actually quite simple. 

If we are on branch `branch` and we would like to change `file.bin`, we can make a commit only if:
  - For every conflicting branch `potential_conflict`:
   `branch` descends from the most recent commits in the history of `potential_conflict` that change `fileA.bin`.

That's it!

The big benefit of this system is that **there is no concept of locks of any sort.** This makes it extremely git-friendly. A lock is simply a change made on any branch, which implicitly locks out other conflicting branches. Releasing a lock is done by simply merging in those changes (or detaching a branch from its dependencies if you'd like to abandon the changes). 

This is a huge benefit because it means it handles arbitrary git actions gracefully. You can cherry-pick, rebase, force-push, etc, without having to make sure that all of the locks are kept up to date and properly organized.

I've actually implemented a prototype of the system in JS, mostly as a sketch to help think through a couple of different cases. You can take a look here: https://kleptine.github.io/git-preemt/ Fair warning though, it's not well documented!

## Open Questions

As I mentioned earlier, this system only functions properly if you have the full commit graph of all current changes local and remote. The main server would need to be kept informed of the commits made on remote repositories. I tend to think that this is unavoidable -- any serious locking system must use the information from all remote repos to see if files are being changed. 

I'm not sure what the best path is to send that information, or keep it up to date. For instance, how large is the full commit graph for the Windows repo? Is it even reasonable to keep the central server updated to all commits made by developers locally? Realistically, we actually don't need the **full** history, only the most recent commit that changed `file.bin` on `potential_conflict` (although, this could be arbitrarily far back in time).

Another open issue is that we really want to be pre-empting modifications of the files in the first place, rather than just at commit time. The sooner we can make the check, the better. This requires, though, that the check is fast enough to perform that it can be done in a couple of milliseconds for the unlocked case.

But the nice thing about this system, is that it makes the problem of networking/synchronization more of an optimization question of keeping a recent view of the global commit graph. I think this is a simpler problem than a generic locking system, and is more in-line with the way GVFS is situated already.

Thoughts?

**Comment by Kleptine** [comment]
Looping in a couple of folks I've seen around github that had some vested interest in general git file locking, in case they had thoughts. 

@sinbad
@NeverNotGlow
@SRombauts

**Comment by sinbad** [comment]
I think you've got some great thoughts going on here. You've come to similar conclusions we did with the system proposed for Git LFS, only a subset of which ever got implemented. I've moved on now and don't need locking myself but the fact that GVFS has more control of the file system layer means you have a much better chance of being able to come up with a solid answer to this, albeit you can never 100% guarantee anything (people can always defy the rules locally when they have access to the OS and thus cause the kinds of problems on the remote that locking was trying to avoid). I tended to look at it in terms of levels of enforcement - with the remote being the hardest rules, and locally the main aim is to prevent accidental issues caused by lack of easy visibility and pro-active notification. 

Everything you say about needing to defy the rules on topic / maintenance branches is true, but more generally there are always going to be cases where people want to break the rules. It's good to try to cater for the most common cases but with the best will in the world you're not going to be able to handle all of them. Another example is experimental branches where the result *may* replace another in future, but we don't know yet. The intention may be to merge it if it works, while at the same time not preventing ongoing smaller refinements to the mainline. So it's like your topic branch example, except that you may need to merge it later (and just blow away the old version). So long as this is with explicit understanding, it's fine. I'd envisaged doing it with a kind of `--force` argument at lock, with a report of the future conflicts (hopefully commit messages would explain the decision if already diverged).

Sorry for the ramble. I think you're very much on the right track with this - when I did the early version in LFS I had all kinds of issues with file permissions because the control we had was not perfect, I hope you find it a bit more manageable with the VFS.

**Comment by Kleptine** [comment]
Appreciate the comments! 

I like the levels of enforcement idea as well. I see the file-system layer as really a best effort system, but I have my doubts about whether it's possible to really fully handle all of the cases.

The "maybe merge" branch situation is a good thought. Perhaps a better way to state the branch dependencies is that drawing an arrow between branches implies "I'd like to check this branch for merge conflicts". So if a branch is sufficiently experimental, you can choose not to check for merge conflicts, or perhaps (I need to think this through a bit), draw an edge between the experimental branch back to master once the branch is solid enough that you feel it should start surfacing conflict errors in master.

I've been doing a bit of work on how to make these queries across all current repos, and I think a 'global commit server' might be a good approach. Essentially as the local repositories change, we push/send the changes to the global server which keeps a single git graph of all commits among all collaborators. It could also maintain the branch references of all contributors (with some sort of namespacing perhaps). I'm not sure if this would scale properly, but it might be a good start, since then the locking queries can be done by making ancestor checks on a single git graph.

One tricky issue is that this server would also have to keep the states of the index of collaborators as well. Perhaps we could hook into the GVFS systems that run when the index changes to forward this information to the synchronization server.

**Comment by sanoursa** [comment]
I think there are a lot of good ideas here, but it is diverging quite far from the purpose of GVFS, which is to make Git scale, but appear as if you're working with normal Git on your normal file system. Right now GVFS does not add any features that Git doesn't have, and does not force you to use any different Git commands than you normally would (outside of `gvfs clone` instead of `git clone`). In its ideal (which is not yet achieved, obviously), GVFS has zero observable behaviors of its own other than perf/scale improvements.

In general, what I would suggest is to first devise a locking scheme that works with normal Git at least one reference Git service. Once that is understood, it could definitely make sense to use GVFS as another enforcement tool, but I still don't think GVFS is in the driver seat on this issue.

I'm going to close this issue because it's not something I would see us pursuing as a _GVFS feature_ unless it first becomes a _Git feature_. I'm more than happy to keep discussing it though, and see if we can land on a good solution.
